cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+#include <cassert>
+#include <cstdlib>
+#include <iostream>
+#include <regex>
+#include <tuple>
+#include <string>
+#include <unordered_map>
+
+#include "btx_cudainterval_callbacks.hpp"
+
+#include "context_manager.hpp"
+
+#include <metababel/metababel.h>
+
+
+struct data_s {
+  // store entry state per thread in order to use in exit callback, conditioned
+  // on the cuResult being a non-error
+  std::unordered_map<hpt_t, int64_t> entry_ts;
+  std::unordered_map<hpt_t, std::string> entry_kernel_name;
+  std::unordered_map<hpt_t, size_t> entry_traffic_size;
+
+  // kernels (CUfunction objects) are "created" before use, at which time we save
+  // the name so at execute time we know the name and can emit an event with the name
+  std::unordered_map<hp_kernel_t, std::string> hp_kernel_to_name;
+
+  // At device operation launch time, save the device the operation was launched on,
+  // keyed by the name, since this is not available in the device profiling events
+  std::unordered_map<hpt_function_name_t, CUdevice> hpt_function_name_to_dev;
+
+  // State for device event profiling
+  //
+  // The flow of events for event profiling is:
+  //
+  // 1. device operation entry (cuMemcpy, cuLaunch, etc)
+  // 2. event_profiling
+  // 3. device operation exit
+  // 4. unknown number of intervening events
+  // ...
+  // 5. event profiling results
+  // 
+  // The information needed to emit a thapi device event is spread out throughout
+  // these calls, so we need to maintain state between them.
+  // 
+  // Since 1 and 2 always happen one after another in a given thread, we just need
+  // to save the function name and launch ts keyed by {hostname, process, thread}.
+  // In (2) we can associate this information with the events, so when the results
+  // are finally available it can be looked up.
+  std::unordered_map<hpt_t, fn_ts_t> hpt_profiled_function_name_and_ts;
+  std::unordered_map<hp_event_t, fn_ts_t> hp_event_to_function_name_and_ts;
+
+  // Encapsulate complex context management state
+  CUDAContextManager context_manager;
+};
+
+using data_t = struct data_s;
+
+void set_entry_ts(data_t *data, hpt_t hpt, int64_t ts) {
+  data->entry_ts[hpt] = ts;
+}
+
+int64_t get_entry_ts(data_t *data, hpt_t hpt) {
+  return data->entry_ts.at(hpt);
```

Comment:
Not sure about an `at` in a performance critical place like this. Bound checking have a place. So maybe we should `ifdef` this guy so that we we are in debug more we use `at` and in release mode we use `[]`? 

I guess I'm just asking for benchmarking, but yeah I'm a little afraid of performance impact